File: python/plugins/solaceMessageProcessor/producer.py
import logging
from solace.messaging.messaging_service import MessagingService
from solace.messaging.resources.topic import Topic
from message_builder import MessageBuilder # Make sure this is correctly imported

logger = logging.getLogger(__name__)


class Producer:
    def __init__(self, messaging_service: MessagingService, topics: list[Topic], message_builder: MessageBuilder):
        self.messaging_service = messaging_service
        self.topics = topics
        self.message_builder = message_builder
        self.publisher = None

    def start(self):
        try:
            self.publisher = self.messaging_service.create_direct_message_publisher_builder().build()
            self.publisher.start()
            logger.info(f"Producer started for topics {[topic.get_name() for topic in self.topics]}")
        except Exception as e:
            logger.error(f"Failed to start publisher: {e}")
            raise

    def publish_messages(self):
        try:
            solace_message_builder = self.messaging_service.message_builder()

            # --- Publish NewOrder ---
            payload_bytes = self.message_builder.build_new_order("ORD001", "AAPL", 100, 150.5)
            # Convert bytes to bytearray to satisfy the IDE's type hint
            message = solace_message_builder.build(bytearray(payload_bytes))
            self.publisher.publish(message, self.topics[0])  # trade/new_order
            logger.info(f"Published message to topic {self.topics[0].get_name()}")

            # --- Publish Cancel ---
            payload_bytes = self.message_builder.build_cancel("ORD001")
            # Convert bytes to bytearray
            message = solace_message_builder.build(bytearray(payload_bytes))
            self.publisher.publish(message, self.topics[1])  # trade/cancel
            logger.info(f"Published message to topic {self.topics[1].get_name()}")

            # --- Publish Amend ---
            payload_bytes = self.message_builder.build_amend("ORD001", 200, 155.75)
            # Convert bytes to bytearray
            message = solace_message_builder.build(bytearray(payload_bytes))
            self.publisher.publish(message, self.topics[2])  # trade/amend
            logger.info(f"Published message to topic {self.topics[2].get_name()}")

        except Exception as e:
            logger.error(f"Failed to publish messages: {e}")
            raise

    def stop(self):
        if self.publisher:
            self.publisher.terminate()
            logger.info(f"Producer stopped for topics {[topic.get_name() for topic in self.topics]}")

File: java/plugins/solaceMessageProcessor/src/main/java/com/applicationsx/trademessageprocessor/Producer.java
package com.applicationsx.trademessageprocessor;

import com.solace.messaging.MessagingService;
import com.solace.messaging.publisher.DirectMessagePublisher;
import com.solace.messaging.publisher.OutboundMessage;
import com.solace.messaging.publisher.OutboundMessageBuilder;
import com.solace.messaging.resources.Topic;
//import com.solace.messaging.message.OutboundMessage;
//import com.solace.messaging.message.OutboundMessageBuilder;

import java.util.List;
import java.util.logging.Logger;

public class Producer {
    private final MessagingService messagingService;
    private final List<Topic> topics;
    private final MessageBuilder messageBuilder;
    private DirectMessagePublisher publisher;
    private static final Logger logger = Logger.getLogger(Producer.class.getName());

    public Producer(MessagingService messagingService, List<Topic> topics, MessageBuilder builder) {
        this.messagingService = messagingService;
        this.topics = topics;
        this.messageBuilder = builder;
    }

    public void start() {
        publisher = messagingService.createDirectMessagePublisherBuilder().build().start();
    }

    public void publishMessages() {
        OutboundMessageBuilder solaceBuilder = messagingService.messageBuilder();

        byte[] newOrderPayload = messageBuilder.buildNewOrder("ORD001", "AAPL", 100, 150.5);
        OutboundMessage msg1 = solaceBuilder.build(newOrderPayload);
        publisher.publish(msg1, topics.get(0));

        byte[] cancelPayload = messageBuilder.buildCancel("ORD001");
        publisher.publish(solaceBuilder.build(cancelPayload), topics.get(1));

        byte[] amendPayload = messageBuilder.buildAmend("ORD001", 200, 155.75);
        publisher.publish(solaceBuilder.build(amendPayload), topics.get(2));
    }

    public void stop() {
        if (publisher != null) publisher.terminate(1000);
    }
}
